### ISSUE TITLE
Will pick cygwin's tar, causing permissions problems

### ISSUE THREAD

**Issue by haarg** [miyagawa/cpanminus#464]
If you have a Win32 system that includes cygwin in its PATH, cpanm will find cygwin's tar.exe and use it for extracting tarballs.  The filesystem semantics it uses don't match how Windows handles them, leading to things like prompting for every file if you try to delete the build files through the GUI.

It might make sense to avoid using Cygwin tools unless `$^O eq 'cygwin'`.

Example `cpanm -v` output:

    C:\Users\Mithaldu>cpanm -v
    cpanm (App::cpanminus) 1.7004 on perl 5.016003 built for MSWin32-x86-multi-thread
    Work directory is C:\Users\Mithaldu/.cpanm/work/1434488587.37360
    You have make c:\Perl\site\bin\dmake.exe
    You have LWP 6.13
    You have C:\cygwin\bin\tar.exe, C:\cygwin\bin\gzip.exe and C:\cygwin\bin\bzip2.exe
    You have C:\cygwin\bin\unzip.exe
    Usage: cpanm [options] Module [...]
     
    Try `cpanm --help` or `man cpanm` for more options.

Example `tar --version` output from cygwin:

    $ tar --version
    tar (GNU tar) 1.28
    Packaged by Cygwin (1.28-1)
    Copyright (C) 2014 Free Software Foundation, Inc.
    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.
    
    Written by John Gilmore and Jay Fenlason.

**Comment by wchristian** [comment]
Maybe relying on this is a good idea:

    C:\Windows\System32>corelist Archive::Tar
    
    Data for 2015-02-14
    Archive::Tar was first released with perl v5.9.3

**Comment by miyagawa** [comment]
Does this work for you?

```
diff --git a/lib/App/cpanminus/script.pm b/lib/App/cpanminus/script.pm
index 997092a..d48c185 100644
--- a/lib/App/cpanminus/script.pm
+++ b/lib/App/cpanminus/script.pm
@@ -16,7 +16,7 @@ use String::ShellQuote ();
 use version ();
 
 use constant WIN32 => $^O eq 'MSWin32';
-use constant BAD_TAR => ($^O eq 'solaris' || $^O eq 'hpux');
+use constant BAD_TAR => (WIN32 || $^O eq 'solaris' || $^O eq 'hpux');
 use constant CAN_SYMLINK => eval { symlink("", ""); 1 };
 
 our $VERSION = $App::cpanminus::VERSION;
```

**Comment by miyagawa** [comment]
Hm, nevermind it already does WIN32 check in maybe_bad_tar. There's some missing combination that it cannot detect when there *is* gzip/bzip.

**Comment by wchristian** [comment]
Does C:\Perl\site\lib\App\cpanminus.pm: 1.7004 have that code already? Maybe my cpanm is too old.

**Comment by miyagawa** [comment]
Yes, 1.7004 has that already. But it only uses it to check if tar has supported uncompression options.

**Comment by miyagawa** [comment]
The options are:

- flip the preference to search for Archive::Tar before searching for tar/gunzip/bunzip2 on Win32
- see if found tar/gunzip/bunzip2 are Cygwin's by looking at its pasth
- see if found tar/gunzip/bunzip2 are Cygwin's by running which.exe and its output (it should return `/usr/bin`)

I think flipping the order is the simplest.

**Comment by haarg** [comment]
Another option is to run tar --version and reject it if it is the Cygwin version and not running under Cygwin.  cpanm already does something along these lines for other tar versions.

I'm not yet advocating for a specific solution to this though.  Still need to give it more thought.
